### DISCUSSION TITLE
I facing difficulty in building executable file

### DISCUSSION THREAD

**Discussion by mohankumar-lookman1** [nodejs/single-executable#95] [Q&A]
The project contains lot of node modules most of them are used for network communication

**while injecting the blob, there are warnings** 
![WhatsApp Image 2024-04-27 at 09 21 48_2e807d97](https://github.com/nodejs/single-executable/assets/144010862/350b4297-bcf6-4fca-8ba4-c0d61952e279)
![WhatsApp Image 2024-04-27 at 09 21 48_6d72c656](https://github.com/nodejs/single-executable/assets/144010862/03745930-3cb3-4cc5-bbb4-2a007174c8c2)

And I am new to this , any help or advices to rectify this is highly appreciable

**Comment by GabenGar** [comment]
Have you read the docs?

**Comment by mohankumar-lookman1** [comment]
yes I do, If something I have missed please guide me through it.

**Comment by rzl** [comment]
If your project is a single file, you can try adding the following code at the top of the file

const { createRequire } = require('node:module');
require = createRequire(__filename); 


======================================

https://nodejs.org/docs/latest/api/single-executable-applications.html#requireid-in-the-injected-main-script-is-not-file-based

require(id) in the injected main script is not file based[#](https://nodejs.org/docs/latest/api/single-executable-applications.html#requireid-in-the-injected-main-script-is-not-file-based)
require() in the injected main script is not the same as the [require()](https://nodejs.org/docs/latest/api/modules.html#requireid) available to modules that are not injected. It also does not have any of the properties that non-injected [require()](https://nodejs.org/docs/latest/api/modules.html#requireid) has except [require.main](https://nodejs.org/docs/latest/api/modules.html#accessing-the-main-module). It can only be used to load built-in modules. Attempting to load a module that can only be found in the file system will throw an error.

Instead of relying on a file based require(), users can bundle their application into a standalone JavaScript file to inject into the executable. This also ensures a more deterministic dependency graph.

However, if a file based require() is still needed, that can also be achieved:

const { createRequire } = require('node:module');
require = createRequire(__filename);

**Reply by mohankumar-lookman1** [reply]
thanks , let me try @rzl it's not a single file but I will include that in entry file.

**Reply by mohankumar-lookman1** [reply]
@rzl  included this lines in entry file(server.js) but not worked

**Reply by rzl** [reply]
Use tools （webpack/rolllup） to package your project into a single file before building executable file

**Reply by mohankumar-lookman1** [reply]
okay I will try that ,  thanks @rzl
